meminfo.c

```diff
@@ -95,6 +116,58 @@ meminfo_t parse_meminfo()
         }
     }
 
+    // If ZFS is in use, compute size of its cache, which is memory-reclaimable.
+    // It currently does not count to `MemAvailable`, like the Linux buffer cache
+    // of other file systems does, see:
+    //   * https://github.com/openzfs/zfs/issues/10255
+    long long zfs_arcstats_bytes = 0;
+    if (arcstats_fd == NULL)
+        arcstats_fd = fopen("/proc/spl/kstat/zfs/arcstats", "r");
+    if (arcstats_fd == NULL && errno != ENOENT) {
+        fatal(106, "could not open /proc/spl/kstat/zfs/arcstats: %s\n", strerror(errno));
+    }
+    if (arcstats_fd != NULL) {
+        rewind(arcstats_fd);
+
+        // Read the file; loop to handle short reads.
+        size_t buf_offset = 0;
+        while (1) {
+            size_t len = fread(arcstats_buf + buf_offset, 1, sizeof(arcstats_buf) - 1 - buf_offset, arcstats_fd);
+            if (ferror(arcstats_fd)) {
+                perror("parse_meminfo: fread() on /proc/spl/kstat/zfs/arcstats failed");
+                fatal(107, "could not read /proc/spl/kstat/zfs/arcstats\n");
+            }
+            buf_offset += len;
+            if (feof(arcstats_fd)) {
+                break;
+            }
+            if (len == 0) {
+                fatal(108, "could not read /proc/spl/kstat/zfs/arcstats: 0 bytes returned\n");
+            }
+        }
+
+        // The format of the `arcstats` file is like this:
+        //     name                            type data
+        //     hits                            4    373259415
+        //     ...
+        //     size                            4    1721339808
+        // We scan for the "size" line and parse the second number.
```

Comment:
Maybe it is possible to use

    get_entry("size                            4", arcstats_buf)

here?

Reply:
I am not sure whether the number of spaces here is guaranteed to be constant. It seems to be "suspiciously enough" to fit the length of all parameters. So far I have not found a specification that discussed this.

Reply:
Hm ok, I agree, too risky.

Reply:
I've added a comment about that.